docs/internals/TLS-SESSIONS.md

```diff
@@ -0,0 +1,156 @@
+wants TLSv1.3, it should not reused a TLSv1.2 session.
+
+### trust anchors
+
+All certificate verification relies on the trust anchors that are used.
+These can be configured in various ways as `CAfile`, `CApath`, `IssuerCert`,
+`CertBlob`, `CAInfoBlob` or `IssuerBlob`. When they are set, the are added
+to the peer_key in the following ways:
+
+ * file and path names are converted to absolute paths. This makes the
+   peer_key independent of the current working directory.
+ * Blobs are hashed using SHA256 and appended in hexadecimal notation.
+
+This makes the peer_key unique for the trust anchors you have configured.
+However, when the files or paths configured change their *content*, this
+may mess peer verification up. Example: you remove a trust anchor from
+a `--cacert` file. Connections reusing a TLS session from a peer no
+longer trusted could then still be successful for a while.
+
+#### client certificate and SRP username/password
+
+Client certificates and SRP username/password are **not** added to the
+peer key. However, when they are used, `:CCERT` or `:SRP-AUTH` is added
+to the key. This prevents mixing of TLS sessions between connections
+that differ here. (There are more checks in the session cache that
+prevent mix up of different client certificates and user info. See below.)
+
+Finally, the peer_key gets the TLS implementation name and version
+added as `:IMPL-<name>/<version>`. This prevents giving session data to
+TLS implementations that might choke on them.
+
+## TLS Session Cache Access
+
+#### Lookups
+
+When a new connection is being established, each SSL connection filter creates
+its own peer_key and calls into the cache. The cache then looks for an entry
+with exactly this peer_key. Peer keys between proxy SSL filters and SSL
+filters talking through a tunnel will be different, as they talk to different
+peers.
+
+If the connection filter wants to use a client certificate or SRP
+authentication, the entry also needs to carry the same or is not a match.
+This works both ways. If the entry carries client cert or SRP auth, a
+connection filter which does not have those will not match either.
+
+On a match, the connection filter gets the session data and feeds that
+to the TLS implementation which, on accepting it, will try to resume it
+for a shorter handshake. In addition, the filter gets the ALPN used
+before and the amount of 0-RTT data that the server announced to be
+willing to accept. The filter can then decide if it wants to attempt
+0-RTT or not.
+
+#### Updates
+
+When a new TLS session is received by a filter, it adds it to the
+cache using its peer_key and SSL configuration. The cache looks for
+a matching entry and, should it find one, replaces the entries data
+with the new session.
+
+Otherwise, it looks for a free entry or purges the oldest or an expired
+entry to make room.
+
+## TLS Session Persistence
+
+Peer key and other information work will when only in memory. When session
```

Comment:
This first sentence could use a rephrase.

Reply:
I rewrote parts. Thanks for looking at my mumblings.